### PR TITLE
JavaKit: check each classPath element exists before starting VM

### PR DESCRIPTION
**Pull request by lhoward**
Partial fix for #136: validate each classPath element by attempting to open it before starting VM. It would be more efficient to stat() it but SystemPackage lacks an abstraction for checking if a file exists, and we don't want to import Foundation.

### CODE REVIEW

**Comment by lhoward** [comment]
Todo: check open-ing a directory is portable (specifically on Windows).

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift [hunk 1]
```diff
@@ -54,6 +56,14 @@ public final class JavaVirtualMachine: @unchecked Sendable {
     // Construct the complete list of VM options.
     var allVMOptions: [String] = []
     if !classPath.isEmpty {
+      for path in classPath {
+        do {
+          let fd = try FileDescriptor.open(path, .readOnly)
+          try? fd.close()
```
Comment (ktoso) [inline Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Can't we do this without a new dependency? I don't love FileManager but it does have such API: `FileManager.default.fileExists(atPath: path)` so we con't have to include another dependency (yet).

https://developer.apple.com/documentation/foundation/filemanager/1410277-fileexists

File: Package.swift [hunk 2]
```diff
@@ -161,7 +162,12 @@ let package = Package(
     ),
     .target(
       name: "JavaKit",
-      dependencies: ["JavaRuntime", "JavaKitMacros", "JavaTypes"],
+      dependencies: [
+        "JavaRuntime",
+        "JavaKitMacros",
+        "JavaTypes",
+        .product(name: "SystemPackage", package: "swift-system")
```
Comment (ktoso) [inline Package.swift]: see comment below about using foundation instead of new dependency

**Comment by lhoward** [comment]
For sure, just thought avoiding Foundation was preferable. This won't pull in ICU?

**Comment by lhoward** [comment]
OK, use FoundationEssentials where possible.

**Comment by lhoward** [comment]
```
  enum JavaKitError: Error {
    case classPathNotFound(String)
  } 
```

maybe we move VMError under this?

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift [hunk 3]
```diff
@@ -12,6 +12,12 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if canImport(FoundationEssentials)
+import FoundationEssentials
+#else
+import Foundation
+#endif
```
Comment (ktoso) [inline Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift:19]: Yup that's the way, thank you

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift [hunk 4]
```diff
@@ -54,6 +60,12 @@ public final class JavaVirtualMachine: @unchecked Sendable {
     // Construct the complete list of VM options.
     var allVMOptions: [String] = []
     if !classPath.isEmpty {
+      let fileManager = FileManager.default
+      for path in classPath {
+        if !fileManager.fileExists(atPath: path) {
+          throw JavaKitError.classPathNotFound(path)
```
Comment (ktoso) [inline Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]:
```suggestion
          throw JavaKitError.classPathSegmentNotFound(notFound: path, classpath: classPath)
```

It's not that the classpath isn't found, just that one of its segments/paths isnt, shall we rename it like this?

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift [hunk 5]
```diff
@@ -268,4 +280,8 @@ extension JavaVirtualMachine {
       }
     }
   }
+
+  enum JavaKitError: Error {
+    case classPathNotFound(String)
```
Comment (ktoso) [inline Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]:
```suggestion
    case classPathSegmentNotFound(notFound: String, classpath: [String])
```

**Review by ktoso — APPROVED** [review]

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift (hunk 5 above)
Comment (lhoward) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Yes, let's just check `segment` is the right word, I'll look at the JDK source to see if it suggests anything (unless you know this for sure?)

Comment (lhoward) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: I think `entry` is used in the JDK.

Comment (ktoso) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Good catch, sounds good!

Comment (lhoward) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: And just to confirm you don't want to make `VMError` a case of `JavaKitError`?

Comment (ktoso) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Either is fine tbh, whichever you prefer. I don’t think we expect people to handle them specifically

**Comment by ktoso** [comment]
Thanks for dealing with the nitpicks :)

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift [hunk 6]
```diff
@@ -268,4 +280,8 @@ extension JavaVirtualMachine {
       }
     }
   }
+
+  enum JavaKitError: Error {
+    case classPathEntryNotFound(entry: String, classPath: String)
```
Comment (ktoso) [inline Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]:
```suggestion
    case classPathEntryNotFound(entry: String, classPath: [String])
```

Comment (ktoso) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Fixing compilation

**Comment by ktoso** [comment]
If you do the "maintainers can edit" on the right I can more easily fix simple stuff like that :) There's a simple compilation error, see proposed fix.

File: Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift (hunk 6 above)
Comment (lhoward) [reply Sources/JavaKit/JavaKitVM/JavaVirtualMachine.swift]: Oh, it was supposed to be `colonSeparatedClassPath` – must have dropped that diff somehow. But I suppose the colon separation is an implementation detail. I'll do as you suggest above.

**Comment by lhoward** [comment]
> If you do the "maintainers can edit" on the right I can more easily fix simple stuff like that :) There's a simple compilation error, see proposed fix.

Hmm, weirdly I can't seem to find this checkbox...

**Comment by ktoso** [comment]
Hmm, don't see something like that? (pic from a different PR that I made) Screenshot 2024-11-05 at 12 32 08

**Comment by ktoso** [comment]
Either way, thank you merged :)

**Comment by lhoward** [comment]
> Hmm, don't see something like that? (pic from a different PR that I made) Screenshot 2024-11-05 at 12 32 08

I think it might be because my fork was in my company repo (PADL) rather than personal. It's a bit hard to parse [1].

[1] https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#

**Comment by ktoso** [comment]
Huh interesting, didn't realize this limitation.
